### ISSUE TITLE
Ticket #13: HuggingFace Adapter Implementation

### ISSUE THREAD

**Issue by MaxenceBouvier** [MaxenceBouvier/form_filler#17]
## Overview
Implement LLM provider adapter using Hugging Face Transformers for intelligent form field matching and value suggestions.

**Priority**: P2 (Medium, Optional)
**Effort**: 5 days
**Dependencies**: #16 (LLM Infrastructure Setup)

## Description
Create a concrete implementation of the `LLMProvider` protocol using Hugging Face Transformers library. This adapter will provide intelligent field matching, value extraction, and form filling suggestions using pre-trained language models. The implementation should be memory-efficient, safe, and work offline with cached models.

## Tasks
- [ ] Create `huggingface_adapter.py` implementing `LLMProvider` protocol
- [ ] Implement model loading with caching and lazy initialization
- [ ] Implement `generate_text` with parameter control
- [ ] Implement `extract_structured_data` using prompts and parsing
- [ ] Implement `suggest_values` for form field suggestions
- [ ] Implement `match_fields` for semantic field matching
- [ ] Add comprehensive safety measures (input/output sanitization)
- [ ] Create `model_manager.py` for model downloading and versioning
- [ ] Add `factory.py` for LLM provider creation
- [ ] Implement memory management and model unloading
- [ ] Create integration tests with small test models
- [ ] Add performance benchmarks

### Core Adapter Implementation
```python
from typing import Any, Optional, List, Tuple, Dict
from dataclasses import dataclass
import torch
from transformers import (
    AutoTokenizer,
    AutoModelForSequenceClassification,
    AutoModelForCausalLM,
    pipeline
)

class HuggingFaceAdapter:
    """HuggingFace Transformers adapter for LLM operations."""

    def __init__(
        self,
        model_name: str = "microsoft/deberta-v3-base",
        device: str = "auto",
        cache_dir: Optional[str] = None,
        max_memory: Optional[int] = None
    ):
        self.model_name = model_name
        self.device = self._determine_device(device)
        self.cache_dir = cache_dir or "~/.cache/form_filler/models"
        self.max_memory = max_memory

        # Lazy loading - models loaded on first use
        self._tokenizer = None
        self._model = None
        self._pipeline = None

    def generate_text(
        self,
        prompt: str,
        max_tokens: int = 100,
        temperature: float = 0.7
    ) -> LLMResponse:
        """Generate text using causal language model."""
        prompt = self._sanitize_input(prompt)

        pipeline = self._get_pipeline("text-generation")
        result = pipeline(
            prompt,
            max_new_tokens=max_tokens,
            temperature=temperature,
            do_sample=temperature > 0,
            pad_token_id=self._tokenizer.eos_token_id
        )

        generated_text = result[0]['generated_text']
        generated_text = self._sanitize_output(generated_text)

        return LLMResponse(
            text=generated_text,
            confidence=result[0].get('score', 1.0),
            metadata={'model': self.model_name}
        )

    def match_fields(
        self,
        source_fields: List[str],
        target_fields: List[str]
    ) -> Dict[str, Tuple[str, float]]:
        """Match fields using semantic similarity."""
        # Use sentence transformers for embeddings
        model = self._get_sentence_transformer()

        source_embeddings = model.encode(source_fields)
        target_embeddings = model.encode(target_fields)

        matches = {}
        for i, source in enumerate(source_fields):
            similarities = torch.cosine_similarity(
                torch.tensor(source_embeddings[i]),
                torch.tensor(target_embeddings)
            )

            best_match_idx = torch.argmax(similarities).item()
            confidence = similarities[best_match_idx].item()

            if confidence > 0.5:  # Threshold
                matches[source] = (target_fields[best_match_idx], confidence)

        return matches

    @property
    def _get_pipeline(self, task: str):
        """Get or create pipeline for task."""
        if self._pipeline is None or self._pipeline.task != task:
            self._pipeline = pipeline(
                task,
                model=self.model_name,
                device=self.device,
                model_kwargs={"cache_dir": self.cache_dir}
            )
        return self._pipeline
```

### Model Manager
```python
# model_manager.py
import hashlib
from pathlib import Path
from typing import Optional, Dict, Any
import json

class ModelManager:
    """Manages model downloading, caching, and versioning."""

    def __init__(self, cache_dir: str = "~/.cache/form_filler/models"):
        self.cache_dir = Path(cache_dir).expanduser()
        self.cache_dir.mkdir(parents=True, exist_ok=True)
        self.registry_file = self.cache_dir / "model_registry.json"

    def download_model(
        self,
        model_name: str,
        force: bool = False
    ) -> Path:
        """Download model if not cached."""
        model_path = self.cache_dir / self._sanitize_model_name(model_name)

        if model_path.exists() and not force:
            return model_path

        # Download using transformers
        from transformers import AutoModel, AutoTokenizer

        AutoModel.from_pretrained(
            model_name,
            cache_dir=str(model_path),
            local_files_only=False
        )
        AutoTokenizer.from_pretrained(
            model_name,
            cache_dir=str(model_path),
            local_files_only=False
        )

        # Update registry
        self._update_registry(model_name, model_path)

        return model_path

    def list_cached_models(self) -> List[Dict[str, Any]]:
        """List all cached models with metadata."""
        if not self.registry_file.exists():
            return []

        with open(self.registry_file) as f:
            return json.load(f)

    def clear_cache(self, model_name: Optional[str] = None):
        """Clear model cache."""
        if model_name:
            model_path = self.cache_dir / self._sanitize_model_name(model_name)
            if model_path.exists():
                shutil.rmtree(model_path)
        else:
            # Clear all
            shutil.rmtree(self.cache_dir)
            self.cache_dir.mkdir(parents=True, exist_ok=True)

    def get_model_size(self, model_name: str) -> int:
        """Get model size in bytes."""
        model_path = self.cache_dir / self._sanitize_model_name(model_name)
        if not model_path.exists():
            return 0

        total_size = 0
        for file in model_path.rglob("*"):
            if file.is_file():
                total_size += file.stat().st_size
        return total_size
```

### Memory Management
```python
class MemoryManager:
    """Manages model memory usage."""

    def __init__(self, max_memory_gb: Optional[float] = None):
        self.max_memory_gb = max_memory_gb or self._get_available_memory()
        self.loaded_models: Dict[str, Any] = {}

    def load_model(self, model_name: str, model_loader):
        """Load model with memory management."""
        required_memory = self._estimate_model_memory(model_name)

        # Check if we need to unload models
        while self._get_used_memory() + required_memory > self.max_memory_gb:
            if not self.loaded_models:
                raise MemoryError(
                    f"Model {model_name} requires {required_memory}GB "
                    f"but only {self.max_memory_gb}GB available"
                )
            self._unload_oldest_model()

        # Load the model
        model = model_loader()
        self.loaded_models[model_name] = {
            'model': model,
            'memory': required_memory,
            'last_used': time.time()
        }
        return model

    def _unload_oldest_model(self):
        """Unload least recently used model."""
        oldest = min(
            self.loaded_models.items(),
            key=lambda x: x[1]['last_used']
        )
        model_name = oldest[0]

        # Clear from memory
        del self.loaded_models[model_name]['model']
        del self.loaded_models[model_name]

        # Force garbage collection
        import gc
        gc.collect()
        torch.cuda.empty_cache() if torch.cuda.is_available() else None
```

## Acceptance Criteria
- [ ] Adapter implements complete `LLMProvider` protocol
- [ ] Model loading is lazy and cached efficiently
- [ ] Memory usage stays within configured limits
- [ ] Outputs are sanitized and validated
- [ ] Graceful degradation when models unavailable
- [ ] Offline mode works with cached models
- [ ] Integration tests pass with small test model
- [ ] Test coverage ≥85% (lower due to external dependency)
- [ ] Performance benchmarks meet targets
- [ ] No memory leaks detected

## Files to Create/Modify
- `src/form_filler/infrastructure/llm/huggingface_adapter.py`
- `src/form_filler/infrastructure/llm/model_manager.py`
- `src/form_filler/infrastructure/llm/memory_manager.py`
- `src/form_filler/infrastructure/llm/factory.py`
- `tests/integration/infrastructure/llm/test_huggingface_adapter.py`
- `tests/integration/infrastructure/llm/test_model_manager.py`

### Unit Tests
```python
def test_adapter_initialization():
    """Test adapter can be created."""
    adapter = HuggingFaceAdapter(model_name="bert-base-uncased")
    assert adapter.model_name == "bert-base-uncased"

def test_input_sanitization():
    """Test input is properly sanitized."""
    adapter = HuggingFaceAdapter()
    malicious = "<script>alert('xss')</script>"
    sanitized = adapter._sanitize_input(malicious)
    assert "<script>" not in sanitized

def test_memory_limits():
    """Test memory limits are respected."""
    adapter = HuggingFaceAdapter(max_memory=1)  # 1GB
    # Test that large models are rejected
    with pytest.raises(MemoryError):
        adapter.load_large_model()
```

### Integration Tests
```python
def test_field_matching():
    """Test semantic field matching."""
    adapter = HuggingFaceAdapter(model_name="sentence-transformers/all-MiniLM-L6-v2")

    source = ["First Name", "Last Name", "Email Address"]
    target = ["given_name", "surname", "email"]

    matches = adapter.match_fields(source, target)

    assert matches["First Name"][0] == "given_name"
    assert matches["Last Name"][0] == "surname"
    assert matches["Email Address"][0] == "email"
    assert all(conf > 0.7 for _, (_, conf) in matches.items())
```

## Performance Targets
- Model loading: < 10 seconds for base models
- Field matching: < 100ms per field pair
- Text generation: < 500ms for 100 tokens
- Memory usage: < 2GB for base models
- Cache hit rate: > 90% after warmup

### Input Sanitization
```python
def _sanitize_input(self, text: str) -> str:
    """Sanitize user input before processing."""
    # Remove HTML/script tags
    text = re.sub(r'<[^>]+>', '', text)

    # Limit length
    max_length = 1000
    text = text[:max_length]

    # Remove control characters
    text = ''.join(char for char in text if char.isprintable())

    return text.strip()
```

### Output Validation
```python
def _validate_output(self, output: Any, expected_type: type) -> bool:
    """Validate LLM output matches expected type."""
    if not isinstance(output, expected_type):
        return False

    # Additional validation based on type
    if expected_type == dict:
        # Check required keys
        # Validate value types
        pass

    return True
```

### For Field Matching
- `sentence-transformers/all-MiniLM-L6-v2` - Fast, accurate
- `sentence-transformers/all-mpnet-base-v2` - More accurate, slower

### For Text Generation
- `microsoft/DialoGPT-small` - Conversational
- `distilgpt2` - General purpose, small

### For Classification
- `microsoft/deberta-v3-base` - High accuracy
- `distilbert-base-uncased` - Fast, lightweight

### Minimal Setup
```python
# Use defaults, small models
adapter = HuggingFaceAdapter()
```

### Production Setup
```python
adapter = HuggingFaceAdapter(
    model_name="sentence-transformers/all-mpnet-base-v2",
    device="cuda" if torch.cuda.is_available() else "cpu",
    cache_dir="/var/cache/form_filler/models",
    max_memory=4  # 4GB limit
)
```

### Offline Setup
```python
# Pre-download models
manager = ModelManager()
manager.download_model("sentence-transformers/all-MiniLM-L6-v2")

# Use offline
adapter = HuggingFaceAdapter(
    model_name="sentence-transformers/all-MiniLM-L6-v2",
    offline_mode=True  # Only use cached
)
```

## Related Issues
- Depends on: #16 (LLM Infrastructure Setup)
- Related to: Application layer use cases that leverage LLM
